test/javadoc-extension-test.js

```diff
@@ -0,0 +1,276 @@
+    }
+    contentCatalog.files.push(entry)
+    return entry
+  }
+
+  const createContentCatalog = () => ({
+    files: [],
+    findBy (criteria) {
+      const criteriaEntries = Object.entries(criteria)
+      const accum = []
+      for (const candidate of this.files) {
+        const candidateSrc = candidate.src
+        if (criteriaEntries.every(([key, val]) => candidateSrc[key] === val)) accum.push(candidate)
+      }
+      return accum
+    },
+  })
+
+  const run = (input = [], opts = {}) => {
+    file.contents = Buffer.from(Array.isArray(input) ? input.join('\n') : input)
+    const context = { config, contentCatalog, file }
+    opts.extension_registry = ext.register(opts.extension_registry || Asciidoctor.Extensions.create(), context)
+    opts.sourcemap = true
+    if (opts.convert) {
+      delete opts.convert
+      return log(Asciidoctor.convert(input, opts))
+    }
+    return log(Asciidoctor.load(input, opts))
+  }
+
+  function log (data) {
+    console.log(data)
+    return data
+  }
+
+  beforeEach(() => {
+    config = {}
+    contentCatalog = createContentCatalog()
+    file = addFile({ relative: 'index.adoc' }, '= Index Page')
+  })
+
+  describe('bootstrap', () => {
+    it('should be able to require extension', () => {
+      expect(ext).to.be.instanceOf(Object)
+      expect(ext.register).to.be.instanceOf(Function)
+    })
+
+    it('should not register to bound extension registry if register function called with no arguments', () => {
+      try {
+        ext.register.call(Asciidoctor.Extensions)
+        const extGroups = Asciidoctor.Extensions.getGroups()
+        const extGroupKeys = Object.keys(extGroups)
+        expect(extGroupKeys).to.be.empty()
+      } finally {
+        Asciidoctor.Extensions.unregisterAll()
+      }
+    })
+
+    it('should not register extension group if context is undefined', () => {
+      const input = []
+      const opts = { extension_registry: ext.register(Asciidoctor.Extensions.create()) }
+      const extensions = Asciidoctor.load(input, opts).getExtensions()
+      expect(extensions).to.be.undefined()
+    })
+
+    it('should be able to call register function exported by extension', () => {
+      const extensions = run().getExtensions()
+      expect(extensions.getInlineMacros()).to.have.lengthOf(1)
+    })
+  })
+
+  describe('javadoc macro', () => {
+    it('should convert using sensible defaults', () => {
+      const input = heredoc`
+        = Page Title
+
+        javadoc:com.example.MyClass[]
+        `
+      const actual = run(input, { convert: true })
+      expect(actual).to.include(
+        '<code class="apiref-hint:com.example.MyClass"><a href="attachment$api/java/com/example/MyClass.html" class="apiref">MyClass</a></code>'
+      )
+    })
+
+    it('should convert with specified location when has javadoc-location attribute', () => {
+      const input = heredoc`
+        = Page Title
+        :javadoc-location: xref:api:java
```

Comment:
Is it intentional to have a `:java` in the location? I think this may break things if/when we want to support module names since the test output also contains the `:` in it. Instead, I'd think this would cause `api` to be a module for the xref but not a part of the href.

Reply:
This might be due to the tests not understanding xrefs fully again. I'm not totally sure what you mean. The `xref:api:java` location is intentional since it should create `xref:api:java/com/example/MyClass.html` links which we [use in Spring Boot](https://github.com/spring-projects/spring-boot/blob/0fafb8d661ab3127f1e5fafc31a844ec8dd168a0/spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/features/json.adoc?plain=1#L37).